clock.go

```diff
@@ -71,29 +64,36 @@ func (clock *Clock) Now() time.Time {
 
 // After is part of the clock.Clock interface.
 func (clock *Clock) After(d time.Duration) <-chan time.Time {
-	defer clock.notifyAlarm()
-	clock.mu.Lock()
-	defer clock.mu.Unlock()
-	notify := make(chan time.Time, 1)
-	if d <= 0 {
-		notify <- clock.now
-	} else {
-		clock.setAlarm(clock.now.Add(d), func() { notify <- clock.now })
-	}
-	return notify
+	return clock.NewTimer(d).Chan()
+}
+
+func (clock *Clock) NewTimer(d time.Duration) clock.Timer {
+	c := make(chan time.Time, 1)
+	return clock.addAlarm(d, c, func() {
+		c <- clock.now
+	})
 }
 
 // AfterFunc is part of the clock.Clock interface.
 func (clock *Clock) AfterFunc(d time.Duration, f func()) clock.Timer {
+	return clock.addAlarm(d, nil, func() {
+		go f()
```

Comment:
I don't *think* so, although I'm open to persuasion. Remember that this implementation is solely for testing purposes, and I it's generally better the less that the tests know about the details of the code (whether the code uses AfterFunc vs After), so the test shouldn't need to change if the code uses AfterFunc (we know when it completes) vs After (it triggers something that happens elsewhere).

And we can't make Stop clean up goroutines because that's not something that the stdlib Timer.Stop method provides - it's something that individual clients can do if they need to.

FWIW AfterFunc is very rarely used in the Juju source code - almost always to just update a local variable or field.